components/core/src/clp/Grep.cpp

```diff
@@ -1063,4 +1044,468 @@ size_t Grep::search(Query const& query, size_t limit, Archive& archive, File& co
+            // Skip strings that begin with an incorrectly unescaped wildcard (e.g., substring
+            // "*text" from string "* \*text *").
+            if (begin_idx > 0 && is_escape[begin_idx - 1]) {
+                continue;
+            }
+            auto possible_substr_types = get_possible_substr_types(
+                    processed_search_string,
+                    begin_idx,
+                    end_idx,
+                    is_greedy_wildcard,
+                    is_non_greedy_wildcard,
+                    is_escape,
+                    lexer
+            );
+            if (possible_substr_types.empty()) {
+                continue;
+            }
+
+            // Use the completed set of variable types for each substr(begin_idx,end_idx) to
+            // construct all possible logtypes for each substr(0,n), for all n.
+            if (begin_idx > 0) {
+                // Handle the case where substr(0,n) is composed of multiple
+                // substr(begin_idx,end_idx).
+                for (auto const& prefix : query_substr_logtypes[begin_idx - 1]) {
+                    for (auto& suffix : possible_substr_types) {
+                        QueryLogtype query_logtype = prefix;
+                        query_logtype.append_logtype(suffix);
+                        query_substr_logtypes[end_idx - 1].insert(query_logtype);
+                    }
+                }
+            } else {
+                // Handle the case where substr(0,n) == substr(begin_idx,end_idx).
+                for (auto& possible_substr_type : possible_substr_types) {
+                    query_substr_logtypes[end_idx - 1].insert(possible_substr_type);
+                }
+            }
+        }
+    }
+    // The last entry of the query_substr_logtypes is the logtypes for the query itself. Convert
+    // this into a vector so we can easily add logtypes when needed.
+    auto& query_logtypes_set = query_substr_logtypes.back();
+    vector<QueryLogtype> query_logtypes;
+    query_logtypes.reserve(query_logtypes_set.size());
+    for (auto it = query_logtypes_set.begin(); it != query_logtypes_set.end();) {
+        query_logtypes.push_back(std::move(query_logtypes_set.extract(it++).value()));
+    }
+    return query_logtypes;
+}
+
+vector<QueryLogtype> Grep::get_possible_substr_types(
+        string& processed_search_string,
+        size_t begin_idx,
+        size_t end_idx,
+        vector<bool>& is_greedy_wildcard,
+        vector<bool>& is_non_greedy_wildcard,
+        vector<bool>& is_escape,
+        ByteLexer& lexer
+) {
+    vector<QueryLogtype> possible_substr_types;
+
+    // Don't allow an isolated wildcard to be considered a variable
+    if (end_idx - 1 == begin_idx && is_greedy_wildcard[begin_idx]) {
+        possible_substr_types.emplace_back('*', "*", false);
+    } else if (end_idx - 1 == begin_idx && is_non_greedy_wildcard[begin_idx]) {
+        possible_substr_types.emplace_back('?', "?", false);
```

Comment:
Can we rewrite these two conditions to be early-returns so that we can reduce the indention of the rest of the function?